Fix race condition in loadUserTimeEntryFromJSON

skel35 opened the pull request:
### 📒 Description
The method `User::loadUserTimeEntryFromJSON` is being called from multiple threads:
- from the thread performing syncing
- from the thread handling WebSocket messages
- from the thread performing Load More
- from the thread performing post-login user data loading

The `Constraint violation` issue can happen when `related.TimeEntries` vector contains time entries with the same ID. 
The way it can happen is when two threads call `User::loadUserTimeEntryFromJSON` at the same time with JSON data of a brand new entry. This way `TimeEntry *model = related.TimeEntryByID(id);` will return `nullptr` for both of the threads and two entries with the same ID would get added to `related.TimeEntries`. Locking the code between getting the time entry by ID and adding a time entry with ID should fix the issue.
This scenario makes sense considering the users reporting these issues usually mention they often create entries with a web app or a mobile app.

### 🕶️ Types of changes
- **Bug fix** (non-breaking change which fixes an issue)

### 👫 Relationships
Closes #3374
Closes #2186

### 🔎 Review hints
I wasn't able to find proper steps to reproduce. See Description on my explanation on how the bug actually happens.
- Review the code itself
- Test the following:
1. time entry creation
2. time entry editing
3. syncing while starting/editing entries on web or mobile
4. clear cache and sync
5. offline mode

skel35 commented on src/user.cc:
```diff
@@ -1064,23 +1064,29 @@ void User::loadUserTimeEntryFromJSON(
-        related.pushBackTimeEntry(model);
+        if (!model) {
+            model = new TimeEntry();
+            model->SetID(id);
```
I've added this one line. `ID` was previously set in `LoadFromJSON()`, but now we want to make sure all entries in `related` have correct IDs before we get out of scope of the lock, so that `related.TimeEntryByID()` returns correct results.

NghiaTranUIT replied:
The code it self is straightforward and this context is locked properly. 

However, I'm not able to reproduce the Constrain issues itself, so I'm not guarantee that this fix could completely fix it or now.

skel35 commented:
Removed the WIP label.
This PR is now ready for review.

NghiaTranUIT commented:
Here is my initial investigation report:

### Why does SQL Constraint error take place? 
<img width="345" alt="Screen Shot 2019-10-22 at 17 31 24" src="https://user-images.githubusercontent.com/5878421/67277774-c8858500-f4f1-11e9-8a1c-89c77378d646.png">
Our database scheme enforces that the `id` and `guid` must be unique. Thus, if we try to insert an TimeEntry, which has `id` matches with the existing TE in the database -> The exception would be thrown. 

Please look back to our `database.cc` where we do `insert` or `update` on `time_entries` table.
https://github.com/toggl/toggldesktop/blob/29b60c16a5d067970efb632d4109a63026969f8f/src/database.cc#L2033-L2035
- This ☝️ func is in charge of saving or updating the TE to the local database.

https://github.com/toggl/toggldesktop/blob/29b60c16a5d067970efb632d4109a63026969f8f/src/database.cc#L2061
- This ☝️ if statement check whether we should do `insert` or `update`.
=> Therefore, it's possible that the race-condition take place here.

- Let check where we update the ID of TimeEntry is:
https://github.com/toggl/toggldesktop/blob/29b60c16a5d067970efb632d4109a63026969f8f/src/time_entry.cc#L414
from https://github.com/toggl/toggldesktop/blob/29b60c16a5d067970efb632d4109a63026969f8f/src/time_entry.cc#L402

- It turns out that this `void TimeEntry::LoadFromJSON(Json::Value data)` is actually executed in many threads and this `id` is not `protected` by any proper locks.
 
<img width="305" alt="Screen Shot 2019-10-22 at 17 39 44" src="https://user-images.githubusercontent.com/5878421/67278326-fa4b1b80-f4f2-11e9-87b8-2fb3418fe0a2.png">

=> My conclusion that by execute `SetID()` from many concurrent threads could be a bottom of the problem.

### Suggestion
1. @skel35 Thank for the first attempt to protect when we setID.
2. Figure it out the solution to protect the `_id` in TimeEntry class.

skel35 commented:
My conclusions after having analyzed all usages of `TimeEntry::LoadFromJSON()`:
1. `BaseModel::LoadFromDataString()` - is only called by an unused `BatchUpdateResult` code. I've created an issue to remove that code #3664. It is out of scope of the current issue.
2. `User::loadUserTimeEntryFromJSON()` - added proper lock protection when calling `TimeEntry::SetID()`.
3. `Context::pushEntries()` - added proper lock protection when calling `TimeEntry::SetID()`. Handled a case when we're trying to add an entry with an already existing ID (very unlikely theoretically possible due to a weird race condition). Added error throwing in the case when the server changes the ID, as it is a backend behavior that we do not expect.

I have moved `TimeEntry::SetID()` out of `LoadFromJSON()` so that it's always protected by a lock.

The PR is now ready for review.
